central/resourcecollection/datastore/datastore_impl.go

```diff
@@ -80,103 +98,170 @@ func (ds *datastoreImpl) initGraph() error {
 
 	// then add edges by batches
 	for i := 0; i < len(ids); i += initBatchSize {
-		var objs []*storage.ResourceCollection
+		var parents []*storage.ResourceCollection
 		if i+initBatchSize < len(ids) {
-			objs, _, err = ds.storage.GetMany(ctx, ids[i:i+initBatchSize])
+			parents, _, err = ds.storage.GetMany(ctx, ids[i:i+initBatchSize])
 		} else {
-			objs, _, err = ds.storage.GetMany(ctx, ids[i:])
+			parents, _, err = ds.storage.GetMany(ctx, ids[i:])
 		}
 		if err != nil {
 			return errors.Wrap(err, "building collection graph")
 		}
-		for _, obj := range objs {
-			for _, peer := range obj.GetEmbeddedCollections() {
-				err = ds.graph.AddEdge(obj.GetId(), peer.GetId())
+		for _, parent := range parents {
+			for _, child := range parent.GetEmbeddedCollections() {
+				err = graph.AddEdge(parent.GetId(), child.GetId())
 				if err != nil {
 					return errors.Wrap(err, "building collection graph")
 				}
 			}
 		}
 	}
+
+	// set the graph object
+	ds.graph = graph
+
 	return nil
 }
 
-func (ds *datastoreImpl) addCollectionToGraph(obj *storage.ResourceCollection) error {
+func (ds *datastoreImpl) addCollectionToGraph(obj *storage.ResourceCollection) (*dag.DAG, error) {
 	ds.initGraphOnce()
-	ds.graphLock.Lock()
-	defer ds.graphLock.Unlock()
 
-	err := ds.graph.AddVertexByID(obj.GetId(), obj.GetName())
+	var err error
+
+	// input validation
+	if err = verifyCollectionObjectNotEmpty(obj); err != nil {
+		return nil, err
+	}
+	if obj.GetId() != "" {
+		return nil, errors.New("invalid argument, added collection must not have a pre-set `id`")
+	}
+
+	// create graph copy to do this operation on
+	graph, err := ds.graph.Copy()
 	if err != nil {
-		return err
+		return nil, err
+	}
+
+	// add vertex
+	id, err := graph.AddVertex(graphEntry{
+		id: uuid.NewV4().String(),
+	})
+	if err != nil {
+		return nil, err
 	}
-	for _, peer := range obj.GetEmbeddedCollections() {
-		err = ds.graph.AddEdge(obj.GetId(), peer.GetId())
+
+	// add edges
+	for _, child := range obj.GetEmbeddedCollections() {
+		err = graph.AddEdge(id, child.GetId())
 		if err != nil {
-			deleteErr := ds.graph.DeleteVertex(obj.GetId())
-			if deleteErr != nil {
-				log.Errorf("Failed to remove collection from internal state object (%v)", deleteErr)
-			}
-			return err
+			return nil, err
 		}
 	}
-	return nil
+
+	// set id for the object and return the new graph
+	obj.Id = id
+	return graph, nil
 }
 
-func (ds *datastoreImpl) deleteCollectionFromGraph(id string) error {
+func (ds *datastoreImpl) updateCollectionInGraph(obj *storage.ResourceCollection) (*dag.DAG, error) {
```

Comment:
nit 1: `updateCollectionInGraphNoLock`
nit 2: add a comment to clarify that this function operates on copy.